packages/keyring-controller/src/KeyringController.ts

```diff
@@ -929,74 +928,7 @@ export class KeyringController extends BaseController<
    * operation completes.
    */
   async persistAllKeyrings(): Promise<boolean> {
-    return this.#withVaultLock(async () => {
-      const { encryptionKey, encryptionSalt } = this.state;
-
-      if (!this.#password && !encryptionKey) {
-        throw new Error(KeyringControllerError.MissingCredentials);
-      }
-
-      const serializedKeyrings = await Promise.all(
-        this.#keyrings.map(async (keyring) => {
-          const [type, data] = await Promise.all([
-            keyring.type,
-            keyring.serialize(),
-          ]);
-          return { type, data };
-        }),
-      );
-
-      serializedKeyrings.push(...this.#unsupportedKeyrings);
-
-      const updatedState: Partial<KeyringControllerState> = {};
-
-      if (this.#cacheEncryptionKey) {
-        assertIsExportableKeyEncryptor(this.#encryptor);
-
-        if (encryptionKey) {
-          const key = await this.#encryptor.importKey(encryptionKey);
-          const vaultJSON = await this.#encryptor.encryptWithKey(
-            key,
-            serializedKeyrings,
-          );
-          vaultJSON.salt = encryptionSalt;
-          updatedState.vault = JSON.stringify(vaultJSON);
-        } else if (this.#password) {
-          const { vault: newVault, exportedKeyString } =
-            await this.#encryptor.encryptWithDetail(
-              this.#password,
-              serializedKeyrings,
-            );
-
-          updatedState.vault = newVault;
-          updatedState.encryptionKey = exportedKeyString;
-        }
-      } else {
-        if (typeof this.#password !== 'string') {
-          throw new TypeError(KeyringControllerError.WrongPasswordType);
-        }
-        updatedState.vault = await this.#encryptor.encrypt(
-          this.#password,
-          serializedKeyrings,
-        );
-      }
-
-      if (!updatedState.vault) {
-        throw new Error(KeyringControllerError.MissingVaultData);
-      }
-
-      const updatedKeyrings = await this.#getUpdatedKeyrings();
-      this.update((state) => {
-        state.vault = updatedState.vault;
-        state.keyrings = updatedKeyrings;
-        if (updatedState.encryptionKey) {
-          state.encryptionKey = updatedState.encryptionKey;
-          state.encryptionSalt = JSON.parse(updatedState.vault as string).salt;
-        }
-      });
-
-      return true;
-    });
+    return this.#withControllerLock(async () => this.#updateVault());
```

Comment:
By moving the logic to `#updateVault`,  we are now able to lock both the general controller operation mutex _and_ the specific vault mutex when calling `persistAllKeyrings` from outside the controller.

This means that:
1. If we persist updates _and_ unlock keyrings in the same "controller operation", persistence and unlock will be mutually exclusive
2. Calling `persistAllKeyrings` from outside the controller is itself a mutually exclusive operation, that needs to wait on other eventual operations running
3. Other public methods that need to persist should now call `this.#updateVault` instead of `this.persistAllKeyrings` to avoid a deadlock